### PR TITLE
Fixed multiple modules that were performing scalar instead of spherical calculations

### LINKED ISSUE #965: pointOnLine is not giving true closest point
Technically closest point should be where the point has right angle with line. Following example shows snapped point is off and actually it is not the closest point on the line.

Jsfiddle example is [here](http://jsfiddle.net/bulutkartal/o2jux5me/).

### LINKED ISSUE #1156: pointToLineDistance strange bahavior
Hi, I've just got strange results from function pointToLineDistance when compared to nearestPointOnLine and to what should be expected. Short explanation code:

```js
let lineCoords = [
	[ 10.964832305908203, 41.004681939880314 ],
	[ 10.977363586425781, 40.99096148527727 ],
	[ 10.983200073242188, 40.97075154073346 ],
	[ 11.02834701538086, 40.98372150040732 ],
	[ 11.02508544921875, 41.00716631272605 ],
	[ 10.994186401367188, 41.01947819666632 ],
	[ 10.964832305908203, 41.004681939880314 ]
];

let line = turf.lineString(lineCoords);

let x0 = 11.02834;
let x1 = 11.02835;
let dx = 0.000001;
for (let i = 0, x = x0; x <= x1; i++, x = x0 + i * dx) {
	let p = turf.point([ x, 41.0 ]);
	let d = turf.pointToLineDistance(p, line, {units : 'meters'});
	let pm = turf.nearestPointOnLine(line, p, {units : 'meters'});
	console.log(turf.getCoord(p), d, pm.properties.dist);
}

```
The result is:
[ 11.02834, 41 ] 188.40111660891222 188.42219043739743
[ 11.028341, 41 ] 188.4845779466462 188.50565183748427
[ 11.028342, 41 ] 188.5680392846129 188.58911323788854
[ 11.028343, 41 ] 188.65150062254605 188.67257463823046
[ 11.028344, 41 ] 188.73496196071352 188.75603603901936
[ 11.028345, 41 ] 188.8184232988473 188.83949743959744
[ 11.028346, 41 ] 188.9018846372153 188.92295884065962
[ 11.028347, 41 ] 188.9853459755496 189.00642024172396
**[ 11.028348, 41 ] 842.5784253401666 189.08988164279026**
[ 11.028349, 41 ] 842.6056956528126 189.17334304432177
[ 11.02835, 41 ] 842.6329734397548 189.2568044457904

There is a threshold longitude for which pointToLineDistance becomes invalid.
Any suggestions?
Thanks

### LINKED ISSUE #1514: nearestPointOnLine broken on vertical line segments
nearestPointOnLine exhibits strange behavior when the line is near-vertical.

For example in this jsfiddle the line goes from [-40,50] to [-40,60], the point is [-40.01,56] and turf returns the nearest point as [-40,~55.3]: https://jsfiddle.net/hdumty60/

I tried playing with the code a bit and the issue went away if I changed nearest-point-on-line to use rhumb destination, not sure if that is an ideal solution though.

### LINKED ISSUE #1878: nearestPointOnLine uses wrong calculation
https://github.com/Turfjs/turf/blob/master/packages/turf-nearest-point-on-line/index.ts

This code is wrong. It is an approximation.

You calculate the azimuth (bearing/direction) between p1 and p2, and from that azimuth, you try to make a perpendicular great circle by changing that azimuth by +180 (or -180, whatever). The problem with this is that, as you move along the path from p1 to p2, your azimuth constantly changes, assuming you are taking the shortest great-circle distance (ie: Haversine/Vincenty). Unless you use a line that assumes the azimuth does not change (ie: occasionally used for moving by sea), your azimuth only applies for p1. p2 has a different azimuth, and so does every single point along that great-circle.

It is a decent approximation of an answer for short distances, but know that this calculation will be way* off if distance(p1, p2) is either long, or if you're close to true north/south pole.

If you do have the proper code for this, I would welcome you let me know.

Thanks

-Phil

### LINKED ISSUE #2023: lineSlice is very imprecise
On turf 6.3.0: lineSlice is very imprecise:

```typescript

      // Start with a line to slice
      const line = {
        "type": "Feature",
        "properties": {
          "index": 0
        },
        "geometry": {
          "type": "LineString",
          "coordinates": [
            [
              3.69140625,
              51.72702815704774
            ],
            [
              -5.3173828125,
              41.60722821271717
            ]
          ]
        }
      }

     // The start point of the line (Just the first coordinate of the line)
     const pointStart = {
        "type": "Feature",
        "properties": {},
        "geometry": {
          "type": "Point",
          "coordinates": [
            3.69140625,
            51.72702815704774
          ]
        }
      }

      // The end point of our slice (obtained by intersecting the line with another line with lineIntersect(line, splitter) )
      const pointEnd = {
        "type": "Feature",
        "properties": {},
        "geometry": {
          "type": "Point",
          "coordinates": [
            0.31936718356317106,
            47.93913163509963
          ]
        }
      }
      
      // Double check that this point is on the line :
      pointToLineDistance(point, segment, { units: "meters" }) < 0.001 ;
      // true
      
      // Now run lineSlice
      lineSlice( pointStart, pointEnd, line )
      // It returns this result:
      {
        "type": "Feature",
        "properties": {
          "index": 0
        },
        "geometry": {
          "type": "LineString",
          "coordinates": [
            // The first coordinate is the start point, normal
            [
              3.69140625,
              51.72702815704774
            ],
            ////////////////////////////////////////////////////////////////////////////////////
            // The second coordinate below is VERY far 
            // from the slice point we asked for, even though we 
            // showed it was on the line
            [
              -0.13132027083960374,
              47.4328630517935
            ]
            // It should be this instead
            // [
            //  0.31936718356317106,
            //  47.93913163509963
            // ]
            ////////////////////////////////////////////////////////////////////////////////////
          ]
        }
      }
```

### LINKED ISSUE #2234: possible bug in lineSlice and nearestPointOnLine
turf version: 6.5.0

I have an application where given an linestring i need a part (slice) of it, from the beginning to some point.  Ultimately I need the distance of the slice of the line.  However, there are points that do not work. Both nearestPointOnLine and lineSlice are failing.  I show

 Below is an example along with the results.

```
let v_cs_ls_p1 = [-64.371054, 25.604782];
let v_cs_ls_p2 = [-19.65732, 45.178866];
let v_cs_ls = turf.lineString([v_cs_ls_p1, v_cs_ls_p2]);

let v_point_on_line = [-58.26491343728603, 28.5528852825411];

let v_point_on_line_adjusted = turf.nearestPointOnLine(v_cs_ls, v_point_on_line, { units: 'meters' });

```

v_point_on_line is on the line, but I call nearestPointOnLine anyway and it returns

```
v_point_on_line_adjusted
{type: 'Feature', properties: {…}, geometry: {…}}
geometry: {type: 'Point', coordinates: Array(2)}
coordinates: (2) [-64.371054, 25.604782]
type: 'Point'
__proto__: Object
properties: {dist: 687546.939020373, index: 0, location: 0}
type: 'Feature'
__proto__: Object
```

as you can see, it thinks the distance is 687546 meters from the line and returns the point which it believes is the nearest which, in this case, is the start point of the line.

The distance should almost be zero (0) if not zero.

to me this seems like a problem, if this is my mistake then fine.  Please show me.

update:  i ran several other tests and this only appears to happen on really long lines.  Long is subjective I know, but long in this case maybe  mean earth factors in.  Not sure.

### LINKED ISSUE #2270: Wrong distance calculated from pointToLineDistance
### Turf version:
6.5.0

### Description of issue:
There seems to be a bug in the calculation of the [minimum distance from a point to a line in turf](https://turfjs.org/docs/#pointToLineDistance).
When calculating the length of a line showing the minimum distance between a point and a line I get a different value than when using the pointToLineDistance function.

image

### Example illustrating the bug:
https://jsfiddle.net/h8c9pvfa/36/

### PR DESCRIPTION
**Pull request by smallsaucepan**
Fixed multiple modules that were performing scalar (2D) calculations when they should have been doing spherical (3D).

Most changes were to nearestPointOnLine. pointToLineDistance now uses nearestPointOnLine for spherical calculations. Flow on corrections impacted lineSlice and nearestPointToLine as well.

Should fix some long standing issues, and be a step towards Turf being internally consistent. For example, a point returned by nearestPointOnLine should have a pointToLineDistance of 0. Or (eventually) return true for booleanPointOnLine.

Resolves #965
Resolves #1156
Resolves #1514
Resolves #1878
Resolves #2023
Resolves #2234
Resolves #2270

- [x] bug fix
- [x] Have read and followed the steps for [preparing a pull request](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#preparing-a-pull-request).

### CODE REVIEW

**Comment by smallsaucepan** [comment]
Btw, benchmark performance tests for nearestPointOnLine were unchanged.

**Review by twelch — APPROVED** [review]
Some minor feedback, otherwise looks good on scan review.  I don't follow the vector math deeply, but trust that you faithfully reimplemented the source.  And I see that all of the added tests demonstrate improvement.

File: packages/turf-nearest-point-on-line/index.ts [hunk 1]
```diff
@@ -162,5 +135,126 @@ function nearestPointOnLine<G extends LineString | MultiLineString>(
...
+  return Math.acos(Math.min(Math.max(theta, -1), 1));
+}
+
+function toRadians(degrees: number) {
+  return degrees * (Math.PI / 180);
+}
+
+function toDegrees(radians: number) {
+  return radians * (180 / Math.PI);
+}
```
[the remark this answers is not in the document]
Comment (smallsaucepan) [reply packages/turf-nearest-point-on-line/index.ts:168]: Can definitely use helpers for toRadians and toDegrees. For others (dot, cross, etc) there's a small third party library I'm going to try using. Having some export / import issues with that though, so can retrofit that once those are sorted.

**Comment by smallsaucepan** [comment]
Thanks @twelch. Will implement all those suggestions and merge shortly.

File: packages/turf-nearest-point-on-line/index.ts [hunk 2]
```diff
@@ -162,5 +135,126 @@ function nearestPointOnLine<G extends LineString | MultiLineString>(
...
+
+function nearestPointOnSegment(
+  posA: Position,
+  posB: Position,
+  posC: Position
+): [Position, boolean, boolean] {
+  // Based heavily on this article on finding cross track distance to an arc:
+  // https://gis.stackexchange.com/questions/209540/projecting-cross-track-distance-on-great-circle
+
+  // Convert lng/lat to spherical coords
```
[the remark this answers is not in the document]
Comment (smallsaucepan) [reply packages/turf-nearest-point-on-line/index.ts:196]: Have expanded on this including a diagram. Misused terminology as well which wouldn't have helped 😬

**Comment by axelvestinsmhi** [comment]
Nice, this is just what i needed!

**Comment by thorseye** [comment]
Just what I was looking for! :) 
@smallsaucepan When can we expect this in a release available on npm?

**Pull request merged**